pkg/transforms/mqttsecret.go

```diff
@@ -0,0 +1,236 @@
+		password:     secrets["password"],
+		keypemblock:  []byte(secrets["clientkey"]),
+		certpemblock: []byte(secrets["clientcert"]),
+		capemblock:   []byte(secrets["cacert"]),
+	}
+
+	return mqttSecrets, nil
+}
+func (sender *MQTTSecretSender) validateSecrets(secrets mqttSecrets) error {
+	caCertPool := x509.NewCertPool()
+	if sender.mqttConfig.AuthMode == AuthModeUsernamePassword {
+		if secrets.username == "" || secrets.password == "" {
+			return errors.New("AuthModeUsernamePassword selected however username or password was not found at vault path")
+		}
+
+	} else if sender.mqttConfig.AuthMode == AuthModeCert {
+		// need both to make a successful connection
+		if len(secrets.keypemblock) <= 0 || len(secrets.certpemblock) <= 0 {
+			return errors.New("AuthModeCert selected however the key or cert PEM block was not found at vault path")
+		}
+	} else if sender.mqttConfig.AuthMode == AuthModeCA && len(secrets.capemblock) <= 0 {
+		return errors.New("AuthModeCA selected however no PEM Block was found at vault path")
+	}
+
+	if len(secrets.capemblock) > 0 {
+		ok := caCertPool.AppendCertsFromPEM([]byte(secrets.capemblock))
+		if !ok {
+			return errors.New("Error parsing CA Certificate")
+		}
+	}
+	return nil
+}
+func (sender *MQTTSecretSender) configureMQTTClient(secrets mqttSecrets) error {
+	var cert tls.Certificate
+	var err error
+	switch sender.mqttConfig.AuthMode {
+	case AuthModeUsernamePassword:
+		sender.opts.SetUsername(secrets.username)
+		sender.opts.SetPassword(secrets.password)
+	case AuthModeCert:
+		cert, err = tls.X509KeyPair(secrets.certpemblock, secrets.keypemblock)
+		if err != nil {
+			return err
+		}
+		fallthrough // continue on to handle CA
+	case AuthModeCA:
+		caCertPool := x509.NewCertPool()
+		ok := caCertPool.AppendCertsFromPEM(secrets.capemblock)
+		if !ok {
+			return errors.New("Error parsing CA PEM block")
+		}
+
+		tlsConfig := &tls.Config{
+			ClientCAs:          caCertPool,
+			InsecureSkipVerify: sender.mqttConfig.SkipCertVerify,
+			Certificates:       []tls.Certificate{cert},
+		}
+		sender.opts.SetTLSConfig(tlsConfig)
+
+	case AuthModeNone:
+		fallthrough
+	default:
+		return nil
+	}
+	return nil
+}
+func (sender *MQTTSecretSender) initializeMQTTClient(edgexcontext *appcontext.Context) error {
+
+	//get the secrets from the vault and populate the struct
+	secrets, err := sender.getSecrets(edgexcontext)
+	if err != nil {
+		return err
+	}
+	//ensure that the authmode selected has the required secret values
+	err = sender.validateSecrets(*secrets)
+	if err != nil {
+		return err
+	}
+	// configure the mqtt client with the retrieved secret values
+	err = sender.configureMQTTClient(*secrets)
+	if err != nil {
+		return err
+	}
+
+	sender.lastRetrieved = time.Now()
+	sender.client = MQTT.NewClient(sender.opts)
+	return nil
+}
+
+func (sender *MQTTSecretSender) connectToBroker(edgexcontext *appcontext.Context, exportData []byte) error {
+	edgexcontext.LoggingClient.Info("Connecting to mqtt server")
```

Comment:
Can we add `for export` so this is not confused with connecting to MQTT for MessageBus. ;-)